simpleviewinc/keg-hub
# Autocomplete component + ReStyle Fix

## Context

* We implemented an `Autocomplete` component ages ago for `tap-zerista`, and I kept kicking a migration down the road 
* There's no more road left so I'm migrating it now
* This should be used in the keg-herkin frontend

## Goal

* Migrate the autocomplete component over, making modifications where necessary to be compatible with our latest keg-components and utilities
* Implement story examples
* Fix a bug in retheme's `reStyle` function in which any update to the styles object would not rerender the component
* Fix a bug in our `Input` component that caused it to lose focus when used as a controlled-component

## Updates

* `repos/keg-components/src/components/form/autocomplete/*.js`
  * root autocomplete component
  * story examples
* `repos/keg-components/src/components/form/scrollable/select/scrollableSelect.js`
  * The scrolling list of possible autocomplete values
* `repos/keg-components/src/components/form/scrollable/item/item.js`
  * the individual row items in the scrollable menu list for autocomplete
* `repos/keg-components/src/components/form/scrollable/select/selectView.js`
  * just a styled view used in the scrollable select to wrap the menu items
* `repos/keg-components/src/components/internal/input.js`
  * removed the Touchable wrapper which was causing the focus-lost bug
  * as far as I can tell, we really don't need the touchable wrapper
  * we aren't using the touchable props like `onPress` prop anywhere in `tap-events-force` or `keg-components`
    * but just in case, I let the `onPress` prop work with `onFocus`
* `repos/keg-components/src/hooks/useAutocompleteItems.js`
  * the hook that does the actual autocompleting work
  * finds the subset of all possible values that overlaps with the user's input
* `repos/keg-components/src/utils/events/getTextFromChangeEvent.js`
  * cross platform helper for getting text from a change event
* `repos/re-theme/src/reStyle/reStyle.js`
  * removed the use of `useObjWithIdentity` in `reStyle`, replacing it with a new hook: `useShallowMemoMerge`
* `repos/re-theme/src/reStyle/reStyleHooks.js`
  * added `useShallowMemoMerge` and unit tests


## Testing

1. `keg dpg run keg-components:autocomplete-component`
2. Navigate to http://components-autocomplete-component.local.kegdev.xyz/?path=/story/components-form-autocomplete--basic
3. Click on the input element, and start typing "Alpine" (or anything really, but Alpine has two branches)
3a. Verify that the menu list appears and is superimposed over the text "Click on an item to select it"
4. Select one of the options in the list
5. Verify that 
    * you see the option you selected in the "Actions" storybook tab
    * what you selected is filled into the input field
6. Now, clear out what you typed, and type in "Lance"
7. Hit the enter key
8. Verify the conditions listed in step 5
9. Clear the field again, then type just "a"
10. Verify you can scroll through the list to see all the items with the letter "a"
11. Verify you can use arrow keys to highlight different items in the list, and that you can press enter to select one
12. Verify that when you click outside of the scrollable list, it hides
